lessons/module-2/introduction-to-typescript.md

```diff
@@ -0,0 +1,320 @@
+### Solution
+1. The solution to converting the variables would look like:
+```ts
+let title: string = "TypeScript Basics";
+let duration: number = 60;
+let isComplete: boolean = false;
+```
+2. The error will be a compile-time error because the type of the variable `score` is not a number.
+</section>
+
+<section class="note">
+### A Note about `any`
+
+TypeScript also supports a type called `any`, which means that a variable can be any type.  This is useful when you want to store a value of any type in a variable.  However, using `any` should be avoided because it disables type-safety.
+
+An example of using `any` would look like:
+```ts
+let id: any = 101;
+id = "E101"; // No error
+```
+</section>
+
+### Union Types
+In addition to type annotation, TypeScript also supports union types.  Union types allow a variable to hold more than one type.  Let's take a look at an example:
+
+```ts
+let id: number | string;
+id = 101; // Valid
+id = "E101"; // Also valid
+```
+
+In this example, the `id` variable can be either a number or a string.  This can be great for when you want to store different types in a single variable.
+
+## TypeScript with Functions
+TypeScript also allows you to specify the types of function parameters and return values.  Let's take a look at an example:
+
+```ts
+function add(a: number, b: number): number {
+  return a + b;
+}
+
+let sum = add(5, 10); // Correct usage
+let result = add("5", 10); // Error: Argument of type 'string' is not assignable to parameter of type 'number'
+```
+
+Note that in the above TypeScript function add, we are defining a function that takes two parameters, `a` and `b`, both of which are of type number. The function is also specified to return a value of type number. If we try to pass in a string for the second parameter, TypeScript will throw an error.
+
+<section class="note">
+### Note: Default Values
+
+TypeScript also supports default values for parameters.  Let's take a look at an example:
+
+```ts
+function greet(name: string, greeting: string = "Hello"): string {
+  return `${greeting}, ${name}!`;
+}
+
+console.log(greet("Alice"));
+console.log(greet("Bob", "Hi"));
+```
+
+1. What do you expect the output to be for `console.log(greet("Alice"));`?
+2. Similarly, what do you expect the output to be for `console.log(greet("Bob", "Hi"));`?
+</section>
+
+<section class="call-to-action">
+### Checks for Understanding
+
+1. Define a function in TypeScript that takes two numbers as arguments and returns their product within a message: `The product of 2 and 3 is 6`. Ensure proper type annotations are used.
+2. Modify the following function to include default parameter values for `taxRate` and `discount`:
+```ts
+function calculateTotalPrice(price: number, taxRate: number, discount: number): number {
+  const taxAmount = price * taxRate;
+  const discountAmount = price * discount;
+  const totalPrice = price + taxAmount - discountAmount;
+  return totalPrice;
+}
+```
+</section>
```

Comment:
Is it worth putting a little more emphasis on the expected return value? Maybe an example where they see what that error would look like if there was a mismatch? i.e

```ts
function multiply(a: number, b: number): number {
  return `The product is ${a * b}.`;
}
```

Reply:
Ohhh good call.  Okay I'll make a quick update to include that and then merge it in.  Thanks Kayla! 🙏